### PR TITLE
New Mesh: WCAtl12to45E2r4

### PR DESCRIPTION
**Pull request by xylar**
Full description to come once we iterate a bit...

### CODE REVIEW

**Comment by xylar** [comment]
@vanroekel, here is the very first draft:
cellWidthGlobal

The transition in resolution in the Arctic needs to change.  My feeling is that the Bering Straight is probably where the transition should happen and the Arctic needs to be part of the RRS30to10 region, but let me know if you want to try something else.

A few more questions:
* Do you want to include ice-shelf cavities around Antarctica?
* Do you want to ease up on the highest resolution around Antarctica?
* Do you want to go with 45 km instead of 60 as the coarsest resolution in the southern Indopacific (as we do in SORRM meshes)?

**Comment by xylar** [comment]
Here's the next draft:
cellWidthGlobal

I'm worried the transition at Bering Straight is too abrupt (from ~10 km to ~60 km resolution over ~1600 km).

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+[main]
+nprocs = 1
+# the file name of the initial condition, possibly after spin-up
+initial_condition = ../../spin_up/simulation/restarts/restart.0001-01-21_00.00.00.nc
+
+[mesh]
+short_name = ${prefix}${min_res}to${max_res}E${e3sm_version}r${mesh_version}
+long_name = ${prefix}${min_res}to${max_res}kmL${levels}E3SMv${e3sm_version}r${mesh_version}
+prefix = WC_SO_Atl
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini:13]: We likely want a different name.  Just `WC`?

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+              ${e3sm_version} with enhanced resolution (${min_res} km) around
+              Antarctica, 45-km resolution in the mid southern latitudes,
+              30-km resolution in a 15-degree band around the equator, 60-km
+              resolution in northern mid latitudes, 30 km in the north
+              Atlantic and 35 km in the Arctic.  This mesh has ${levels}
+              vertical levels and includes cavities under the ice shelves
+              around Antarctica. The mesh is described in the following pull
+              request: https://github.com/MPAS-Dev/compass/pull/37
+e3sm_version = 2
+mesh_version = x
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini:24]: `r4`?

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+prefix = WC_SO_Atl
+description = MPAS Southern Ocean regionally refined mesh for E3SM version
+              ${e3sm_version} with enhanced resolution (${min_res} km) around
+              Antarctica, 45-km resolution in the mid southern latitudes,
+              30-km resolution in a 15-degree band around the equator, 60-km
+              resolution in northern mid latitudes, 30 km in the north
+              Atlantic and 35 km in the Arctic.  This mesh has ${levels}
+              vertical levels and includes cavities under the ice shelves
+              around Antarctica. The mesh is described in the following pull
+              request: https://github.com/MPAS-Dev/compass/pull/37
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini:22]: Needs revision!

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+e3sm_version = 2
+mesh_version = x
+creation_date = autodetect
+min_res = 12
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini:26]: Update once we settle on the design.

File: ocean/global_ocean/WC_SO_Atl10/init/config_culled_mesh.xml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<config case="culled_mesh">
+
+	<add_link source="../base_mesh/base_mesh.nc" dest="base_mesh.nc"/>
+	<add_link source_path="script_configuration_dir" source="scripts/cull_mesh.py" dest="cull_mesh.py"/>
+
+	<run_script name="run.py">
+		<step executable="./cull_mesh.py">
+			<argument flag="--with_critical_passages"></argument>
+            <argument flag="--with_cavities"></argument>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_culled_mesh.xml:10]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_driver_init.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	</case>
+	<case name="culled_mesh">
+		<step executable="./run.py" quiet="true" pre_message=" * Culling land cells from 2D global mesh, including ice shelf cavities..." post_message="     complete!  Created file:  culled_mesh/culled_mesh.nc"/>
+	</case>
+	<case name="initial_state">
+		<step executable="./run.py" quiet="true" pre_message=" * Initializing ocean state with bathymetry and tracers..." post_message="     complete!  Created file:  initial_state/initial_state.nc"/>
+	</case>
+	<case name="ssh_adjustment">
+		<step executable="./run.py" quiet="true" pre_message=" * Adjusting SSH for ice shelf cavities..." post_message="     complete!  Created file:  ssh_adjustment/init.nc"/>
+	</case>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_driver_init.xml:13]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_driver_init.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	<case name="test">
+		<step executable="./run.py" quiet="true" pre_message=" * Quick test run with ice shelf cavities..." post_message="     complete!  Created file:  test/land_ice_fluxes.nc"/>
+	</case>
+	<validation>
+		<compare_fields file1="test/output.nc">
+			<template file="prognostic_comparison.xml" path_base="script_core_dir" path="templates/validations"/>
+		</compare_fields>
+		<compare_fields file1="test/land_ice_fluxes.nc">
+			<template file="land_ice_flux_comparison.xml" path_base="script_core_dir" path="templates/validations"/>
+		</compare_fields>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_driver_init.xml:23]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_initial_state.xml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	<add_link source_path="initial_condition_database" source="PotentialTemperature.01.filled.60levels.PHC.151106.nc" dest="temperature.nc"/>
+	<add_link source_path="initial_condition_database" source="Salinity.01.filled.60levels.PHC.151106.nc" dest="salinity.nc"/>
+	<add_link source_path="initial_condition_database" source="windStress.ncep_1958-2000avg.interp3600x2431.151106.nc" dest="wind_stress.nc"/>
+	<add_link source_path="bathymetry_database" source="BedMachineAntarctica_and_GEBCO_2019_0.05_degree.200128.nc" dest="topography.nc"/>
+	<add_link source_path="initial_condition_database" source="chlorophyllA_monthly_averages_1deg.151201.nc" dest="swData.nc"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="template_initial_state.xml" path_base="script_configuration_dir"/>
+		<template file="template_critical_passages.xml" path_base="script_core_dir" path="global_ocean"/>
+		<template file="template_init_with_land_ice.xml" path_base="script_configuration_dir"/>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_initial_state.xml:38]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_initial_state.xml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	</get_file>
+
+	<get_file dest_path="initial_condition_database" file_name="chlorophyllA_monthly_averages_1deg.151201.nc">
+		<mirror protocol="wget" url="https://web.lcrc.anl.gov/public/e3sm/mpas_standalonedata/mpas-ocean/initial_condition_database/"/>
+	</get_file>
+
+	<add_link source="../culled_mesh/culled_mesh.nc" dest="mesh.nc"/>
+	<add_link source="../culled_mesh/culled_graph.info" dest="graph.info"/>
+	<add_link source="../culled_mesh/critical_passages_mask_final.nc" dest="critical_passages.nc"/>
+	<add_link source="../culled_mesh/land_ice_mask.nc" dest="land_ice_mask.nc"/>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_initial_state.xml:26]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_ssh_adjustment.xml [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0"?>
+<config case="ssh_adjustment">
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_ssh_adjustment.xml:2]: Remove if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/init/config_test.xml [hunk 11]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0"?>
+<config case="test">
+	<add_link source="../ssh_adjustment/init20.nc" dest="init.nc"/>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/init/config_test.xml:3]: Fix if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/spin_up/config_driver.xml [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+		<step executable="./run.py" quiet="true" pre_message=" * Running prep_spin_up3" post_message=" - Complete"/>
+	</case>
+	<case name="simulation">
+		<step executable="./run.py" quiet="true" pre_message=" * Running simulation" post_message=" - Complete"/>
+	</case>
+	<validation>
+		<compare_fields file1="simulation/output.nc">
+			<template file="prognostic_comparison.xml" path_base="script_core_dir" path="templates/validations"/>
+		</compare_fields>
+	</validation>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/spin_up/config_driver.xml:18]: Update based on @vanroekel and @lconlon's new strategy.

File: ocean/global_ocean/WC_SO_Atl10/template_forward.xml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+<template>
+	<namelist>
+		<option name="config_dt">'00:10:00'</option>
+		<option name="config_btr_dt">'00:00:20'</option>
+		<option name="config_pio_num_iotasks">36</option>
+		<option name="config_pio_stride">36</option>
+		<option name="config_run_duration">'0000_01:00:00'</option>
+		<option name="config_mom_del4">1.5e10</option>
+		<option name="config_hmix_scaleWithMesh">.true.</option>
+		<option name="config_use_GM">.false.</option>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/template_forward.xml:10]: @vanroekel, GM is now off.

File: ocean/global_ocean/WC_SO_Atl10/template_forward.xml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+		<option name="config_check_ssh_consistency">.false.</option>
+		<option name="config_land_ice_flux_mode">'standalone'</option>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/template_forward.xml:12]: Change if no ice-shelf cavities

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<template>
+	<namelist>
+		<option name="config_global_ocean_topography_smooth_iterations">10</option>
+		<option name="config_global_ocean_topography_smooth_weight">0.92</option>
```
Comment (xylar) [inline ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml:5]: Not sure we want topography smoothing.

**Comment by vanroekel** [comment]
Thanks for starting this so quickly @xylar!  It is looking pretty great already.  A couple replies

1) I agree the Bering Strait transition looks pretty rough.  What if we do two things, (1) make the max resolution 45 km as you said instead of 60 and (2) come back to 12 km in the arctic instead of 10km.

2) I don't really want to coarsen Antarctic resolution.  The transports through the S. Ocean in SORRM look really good, I'd hate to mess with that much.

3) I'd like to go no ice cavities for one version of the mesh for WC.

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml (hunk 15 above)
Comment (vanroekel) [reply ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml:5]: My thinking is we probably do, but for consistency, I'd like to stick with whatever we did for the WC14 mesh.

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini (hunk 1 above)
Comment (vanroekel) [reply ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini:13]: I'm thinking WC_Atl would be a reasonable choice.  We may want Atl as just having the resolution min/max could be misleading.

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml (hunk 15 above)
Comment (xylar) [reply ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml:5]: Okay, I'll check on that.

**Comment by xylar** [comment]
@vanroekel, that all sounds good.  I'll make revisions and start a new build in Anvil soon.

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml (hunk 15 above)
Comment (xylar) [reply ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml]: WC14 didn't include any smoothing, but the SORRM mesh did.  For now, I have taken out smoothing to be consistent with WC14, but let me know if you'd prefer to be consistent with SORRM.

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini (hunk 1 above)
Comment (xylar) [reply ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini]: Done!

**Comment by xylar** [comment]
@vanroekel, here's the latest res plot.  Let me know ASAP if you want something changed.

cellWidthGlobal

**Comment by xylar** [comment]
I'm going to generate the culled mesh here on Anvil:
/lcrc/group/e3sm/ac.xylar/compass/test_WC_SO_Atl/WC_Atl12/ocean/global_ocean/WC_Atl12/

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml (hunk 15 above)
Comment (vanroekel) [reply ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml]: After pondering this a bit more, I'd say we put our best configuration forward and I think that would include smoothing.  Let's switch to consistent with SORRM.  Sorry for changing again.

**Comment by vanroekel** [comment]
the new plot looks great!

File: ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini (hunk 1 above)
Comment (xylar) [reply ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini]: I think WCAtl (without any underscores) would be better.  I think I recall there were problems with punctuation in mesh names (possibly including underscores).

Comment (vanroekel) [reply ocean/global_ocean/WC_SO_Atl10/files_for_e3sm/config_files_for_e3sm.ini]: WCAtl sounds good to me.

File: ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml (hunk 15 above)
Comment (xylar) [reply ocean/global_ocean/WC_SO_Atl10/template_initial_state.xml]: No problem, that's why I asked.

**Comment by xylar** [comment]
Okay, I'm going to let it get through culling the mesh with the setup as it was (because no recent changes affect the culled mesh).  Then, I'll start over with the recent changes.

In the meantime, @vanroekel and @lconlon, could you look over the namelist options for init mode and the whole setup for "spinup" so we can get that plumbing in place later today?

**Comment by milenaveneziani** [comment]
@xylar , @vanroekel: I don't think the transition at Bering St was changed in the latest iteration, is that correct? It does seem a bit too sharp to me at the moment.

**Comment by milenaveneziani** [comment]
Just for reference, in the ARRM meshes, we start the transition in the Pacific at ~45N.

**Comment by vanroekel** [comment]
@milenaveneziani I thought so too, but the color bar max was changed so the transition width is the same but is less sharp

**Comment by vanroekel** [comment]
@xylar sure I will take a look at the settings today

**Comment by xylar** [comment]
> Just for reference, in the ARRM meshes, we start the transition in the Pacific at ~45N.

It's a little tricky but I could try to make the transition wider.  (It's tricky because the transition is for the whole shape defining the Atlantic/Arctic, not just the Bering Straight.)  But part of  my concern is that this mesh won't be appreciably cheaper than a pure RRS30to12 if we push the higher resolution too far into the Indopacific.  So maybe there's a compromise where the transition is a bit wider but not all the way to 45 N.

**Comment by vanroekel** [comment]
I would prefer to leave the transition as is in the figure.  It seems unlikely to me that modifying the transition position or width here will have any appreciable affect on WC simulations.  I agree with @xylar that moving this transition out would make this about the same cost as an eddy resolving mesh everywhere, which we are trying to avoid.

**Comment by vanroekel** [comment]
a similar transition is used in WC14, see figure here

https://github.com/MPAS-Dev/MPAS-Model/pull/628

I believe the transition width we use is from Kristin H.'s paper, I'm pretty comfortable with the transition as is.

**Comment by milenaveneziani** [comment]
ok, I didn't realize that was the same as in WC14. I'll stop bugging you ;)

**Comment by vanroekel** [comment]
No worries, thanks for asking the questions and the review!

**Comment by xylar** [comment]
@vanroekel, culling this mesh is going to take many, many hours.  I have a 4-hour job just for this on Anvil but I think it's going to time out.  I'm going to put in a 10-hour job now to make sure it's done by the time I get up tomorrow.

**Comment by xylar** [comment]
@vanroekel, one more update.

1. Culling seems to be almost done, actually.  
2. I'll see if I can get an initial state.  But took out a template that needed to be there, so need to fix that.
3. In the process, I realized that the SORRM mesh *doesn't* have topography smoothing.  I'll make completely sure but the template I though was adding topographic smoothing is, in fact, the one that sets a bunch of other parameters (see above).

**Comment by darincomeau** [comment]
@xylar do you have the number of horizontal cells (or do you need the culled mesh for that)? For reference, the RRS30to10v3 has 1,445,361.

**Comment by xylar** [comment]
The new version will show up here.  I just submitted an 8-hour job to create it.  Fingers crossed:
```
/home/ac.xylar/e3sm/ac.xylar/compass/test_WCAtl12/ocean/global_ocean/WCAtl12
```

**Comment by xylar** [comment]
> @xylar do you have the number of horizontal cells (or do you need the culled mesh for that)?

Yes, I need the culled mesh for that but I have it:
```
	nCells = 851664 ;
	nEdges = 2576825 ;
	nVertices = 1724606 ;
```

**Comment by xylar** [comment]
For comparison, SORRMr4 has:
```
nCells = 569915 ;
nEdges = 1723935 ;
nVertices = 1153614 ;
```
and my proposed SORRMr5 mesh would have:
```
nCells = 609013 ;
nEdges = 1841499 ;
nVertices = 1232053 ;
```

**Comment by xylar** [comment]
If anyone wants to just look at the culled mesh, it's at:
```
/home/ac.xylar/e3sm/ac.xylar/compass/test_WC_SO_Atl/WC_Atl12/ocean/global_ocean/WC_Atl12/init/culled_mesh/culled_mesh.nc
```
The same thing should eventually end up in the newer directory I listed above:
```
/home/ac.xylar/e3sm/ac.xylar/compass/test_WCAtl12/ocean/global_ocean/WCAtl12
```

**Comment by vanroekel** [comment]
This is great progress.  Thanks @xylar!  I'm happy to see this come in around 2x WC14.  Hopefully this helps WC issues!

**Comment by xylar** [comment]
@vanroekel, the mesh and initial condition are now available on Anvil.  I had to fix the namelist options related to the topography (because we are using the GEBCO/BedMachine topography, which has different variable names).

The short `test` run completed successfully.  I didn't try to run the `spinup` because I think you and @lconlon are going to completely change that anyway.

Please push to this branch or make suggestions for modifying `spinup`, and I'll run the rest to get the files that @darincomeau and @jonbob need to get this ready in E3SM.

**Comment by vanroekel** [comment]
@xylar I've modified the spin up.  WC never had a chance to switch to the new spin up, so I think it is best to do spin up as we have done but set parameters more closely aligned with target settings for E3SM, I've done so here.

**Comment by xylar** [comment]
Thanks @vanroekel.  I tried spinning up but we seem to be getting CFL violations within the first 10 hours.  I have a meeting but will try to play with parameters in an hour or so.

**Comment by vanroekel** [comment]
Thanks @xylar.  Please let me know if you'd like any help tinkering to get this stable.  Happy to jump in if it would be useful.

**Comment by xylar** [comment]
It seems like it's just a time step that's too long.  Should be easy to fix.

**Comment by xylar** [comment]
Reducing the time step in the first stage seems to have done it.  Probably overkill but I don't feel like playing around with it.

I'm working on the final stage (the files for E3SM and MPAS-Analysis).

**Comment by xylar** [comment]
@vanroekel, @darincomeau and @jonbob, the files are in place, as I just emailed you:
```
/lcrc/group/e3sm/ac.xylar/compass/test_WCAtl12/ocean/global_ocean/WCAtl12/files_for_e3sm/files_for_e3sm/assembled_files_for_upload/inputdata/
```

@vanroekel, let me know when you think this mesh revision is "finalized" and I'll merge this PR.

**Review by vanroekel — APPROVED** [review]
Approving based on visual inspection, figures posted to the PR and testing from @darincomeau

**Comment by vanroekel** [comment]
sorry for the letting this one slip @xylar I think we can finalize this and integrate to compass.  Thanks again for the quick and great work!

**Comment by xylar** [comment]
Thanks @vanroekel.  Definitely no rush but it's also good to get this merged so we don't forget.

**Pull request merged**